// ember-handlebars-brunch/lib/ember-handlebars-compiler.js
(function() {

  module.exports = (function() {
    var context, element, emberjs, emberjsPath, fs, handlebarsjs, handlebarsjsPath, jQuery, sandbox, sysPath, vm;
    fs = require('fs');
    vm = require('vm');
    sysPath = require('path');
    handlebarsjsPath = sysPath.join(__dirname, '..', 'vendor', 'handlebars-1.0.0.beta.6.js');
    emberjsPath = sysPath.join(__dirname, '..', 'vendor', 'ember.js');
    handlebarsjs = fs.readFileSync(handlebarsjsPath, 'utf8');
    emberjs = fs.readFileSync(emberjsPath, 'utf8');
    jQuery = function() {
      return jQuery;
    };
    jQuery.ready = function() {
      return jQuery;
    };
    jQuery.inArray = function() {
      return jQuery;
    };
    jQuery.jquery = "1.7.1";
    element = {
      firstChild: function() {
        return element;
      },
      innerHTML: function() {
        return element;
      }
    };
    sandbox = {
      document: {
        createRange: false,
        createElement: function() {
          return element;
        }
      },
      console: console,
      jQuery: jQuery,
      $: jQuery,
      template: null,
      templatejs: null
    };
    sandbox.window = sandbox;
    context = vm.createContext(sandbox);
    vm.runInContext(handlebarsjs, context, 'handlebars.js');
    vm.runInContext(emberjs, context, 'ember.js');
    return function(templateData) {
      context.template = templateData;
      vm.runInContext('templatejs = Ember.Handlebars.precompile(template).toString();', context);
      return context.templatejs;
    };
  })();

}).call(this);
