// ember-handlebars-brunch/lib/index.js
(function() {
  var EmberHandlebarsCompiler, compileHBS, sysPath;

  sysPath = require('path');

  compileHBS = require('./ember-handlebars-compiler');

  module.exports = EmberHandlebarsCompiler = (function() {

    EmberHandlebarsCompiler.prototype.brunchPlugin = true;

    EmberHandlebarsCompiler.prototype.type = 'template';

    EmberHandlebarsCompiler.prototype.extension = 'hbs';

    EmberHandlebarsCompiler.prototype.precompile = false;

    function EmberHandlebarsCompiler(config) {
      this.config = config;
      if (this.config.files.templates.precompile === true) this.precompile = true;
      null;
    }

    EmberHandlebarsCompiler.prototype.compile = function(data, path, callback) {
      var content, error, result;
      try {
        if (this.precompile === true) {
          content = compileHBS(data.toString());
          return result = "\nEmber.TEMPLATES[module.id] = Ember.Handlebars.template(" + content + ");\n module.exports = module.id;";
        } else {
          content = JSON.stringify(data.toString());
          return result = "\nEmber.TEMPLATES[module.id] = Ember.Handlebars.compile(" + content + ");\n module.exports = module.id;";
        }
      } catch (err) {
        return error = err;
      } finally {
        callback(error, result);
      }
    };

    return EmberHandlebarsCompiler;

  })();

}).call(this);
