// mocha/lib-cov/reporters/index.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['reporters/index.js']) {
  _$jscoverage['reporters/index.js'] = [];
  _$jscoverage['reporters/index.js'][2] = 0;
  _$jscoverage['reporters/index.js'][3] = 0;
  _$jscoverage['reporters/index.js'][4] = 0;
  _$jscoverage['reporters/index.js'][5] = 0;
  _$jscoverage['reporters/index.js'][6] = 0;
  _$jscoverage['reporters/index.js'][7] = 0;
  _$jscoverage['reporters/index.js'][8] = 0;
  _$jscoverage['reporters/index.js'][9] = 0;
  _$jscoverage['reporters/index.js'][10] = 0;
  _$jscoverage['reporters/index.js'][11] = 0;
  _$jscoverage['reporters/index.js'][12] = 0;
  _$jscoverage['reporters/index.js'][13] = 0;
  _$jscoverage['reporters/index.js'][14] = 0;
  _$jscoverage['reporters/index.js'][15] = 0;
}
_$jscoverage['reporters/index.js'][2]++;
exports.Base = require("./base");
_$jscoverage['reporters/index.js'][3]++;
exports.Dot = require("./dot");
_$jscoverage['reporters/index.js'][4]++;
exports.Doc = require("./doc");
_$jscoverage['reporters/index.js'][5]++;
exports.TAP = require("./tap");
_$jscoverage['reporters/index.js'][6]++;
exports.JSON = require("./json");
_$jscoverage['reporters/index.js'][7]++;
exports.HTML = require("./html");
_$jscoverage['reporters/index.js'][8]++;
exports.List = require("./list");
_$jscoverage['reporters/index.js'][9]++;
exports.Spec = require("./spec");
_$jscoverage['reporters/index.js'][10]++;
exports.Progress = require("./progress");
_$jscoverage['reporters/index.js'][11]++;
exports.Landing = require("./landing");
_$jscoverage['reporters/index.js'][12]++;
exports.JSONCov = require("./json-cov");
_$jscoverage['reporters/index.js'][13]++;
exports.HTMLCov = require("./html-cov");
_$jscoverage['reporters/index.js'][14]++;
exports.JSONStream = require("./json-stream");
_$jscoverage['reporters/index.js'][15]++;
exports.XUnit = require("./xunit");
_$jscoverage['reporters/index.js'].source = ["","exports.Base = require('./base');","exports.Dot = require('./dot');","exports.Doc = require('./doc');","exports.TAP = require('./tap');","exports.JSON = require('./json');","exports.HTML = require('./html');","exports.List = require('./list');","exports.Spec = require('./spec');","exports.Progress = require('./progress');","exports.Landing = require('./landing');","exports.JSONCov = require('./json-cov');","exports.HTMLCov = require('./html-cov');","exports.JSONStream = require('./json-stream');","exports.XUnit = require('./xunit')"];
