// @angular-devkit/schematics/test/schematic-test-runner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
const schematics_1 = require("@angular-devkit/schematics");
const tools_1 = require("@angular-devkit/schematics/tools");
const json_schema_1 = require("@ngtools/json-schema");
const Observable_1 = require("rxjs/Observable");
class SchematicTestRunner {
    constructor(collectionName) {
        this.collectionName = collectionName;
        this.prepareCollection();
    }
    prepareCollection() {
        this.engineHost = new tools_1.NodeModulesEngineHost();
        this.engine = new schematics_1.SchematicEngine(this.engineHost);
        this.engineHost.registerOptionsTransform((schematic, opts) => {
            if (schematic.schema && schematic.schemaJson) {
                const SchemaMetaClass = json_schema_1.SchemaClassFactory(schematic.schemaJson);
                const schemaClass = new SchemaMetaClass(opts);
                return schemaClass.$$root();
            }
            return opts;
        });
        this.collection = this.engine.createCollection(this.collectionName);
    }
    runSchematicAsync(schematicName, opts, tree) {
        const schematic = this.collection.createSchematic(schematicName);
        const host = Observable_1.Observable.of(tree || new schematics_1.VirtualTree);
        return schematic.call(opts || {}, host);
    }
    runSchematic(schematicName, opts, tree) {
        const schematic = this.collection.createSchematic(schematicName);
        let result = null;
        const host = Observable_1.Observable.of(tree || new schematics_1.VirtualTree);
        schematic.call(opts || {}, host)
            .subscribe(t => result = t);
        if (result === null) {
            throw new Error('Schematic is async, please use runSchematicAsync');
        }
        return result;
    }
}
exports.SchematicTestRunner = SchematicTestRunner;
